### PR TITLE
cereal: add patch for CVE-2020-11105

### PR DESCRIPTION
**Pull request by talw**
###### Motivation for this change

Fixes: https://nvd.nist.gov/vuln/detail/CVE-2020-11105

A patch to 1.3.0 exists upstream (1.3.1 not released yet):
https://github.com/USCiLab/cereal/pull/667/commits/3e8b9cab82dc4a596dc3bcfe7437946395ad782b

This is the first time I am PRing this kind of commit so please let me know if
I am missing something.

Cereal 1.3.0 seems to exist on 20.09 and 20.03 as well, should this be backported to both of them?
I only ran `nixpkgs-review` on master for this PR.
On the off chance that some reverse dependencies fail to build on 20.09 or 20.03 is the backport for that specific release aborted?

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 121574` at d720aa38 run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>2 packages built successfully:</summary>
  <ul>
    <li>cereal</li>
    <li>domoticz</li>
  </ul>
</details>
<details>
  <summary>2 suggestions:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md">maintainers-missing</a>

Package does not have a maintainer. Consider adding yourself?

Near pkgs/development/libraries/cereal/default.nix:30:3:
```
   |
30 |   meta = with lib; {
   |   ^
```
</li>
    <li>error: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-flags-array.md">no-flags-array</a>

`cmakeFlagsArray` is only intended to be used in Bash, not as a Nix attribute.

Near pkgs/development/libraries/cereal/default.nix:28:3:
```
   |
28 |   cmakeFlagsArray = [ "-DJUST_INSTALL_CEREAL=yes" ];
   |   ^
```
</li>
  </ul>
</details>

**Comment by talw** [comment]
> Result of `nixpkgs-review pr 121574` at [d720aa3](https://github.com/NixOS/nixpkgs/commit/d720aa3809c07804289481f7e2e85640b4192991) run on aarch64-linux [1](https://github.com/Mic92/nixpkgs-review)
> 
> 2 packages built successfully:
> 2 suggestions:
> * warning: [maintainers-missing](https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/maintainers-missing.md)
>   Package does not have a maintainer. Consider adding yourself?
>   Near pkgs/development/libraries/cereal/default.nix:30:3:
>   ```
>      |
>   30 |   meta = with lib; {
>      |   ^
>   ```
> * error: [no-flags-array](https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/no-flags-array.md)
>   `cmakeFlagsArray` is only intended to be used in Bash, not as a Nix attribute.
>   Near pkgs/development/libraries/cereal/default.nix:28:3:
>   ```
>      |
>   28 |   cmakeFlagsArray = [ "-DJUST_INSTALL_CEREAL=yes" ];
>      |   ^
>   ```

Regarding `no-flags-array`, I changed `cmakeFlagsArray` to `cmakeFlags`.

Regarding `maintainers-missing` warning, I never used this library before, nor
am I programming in C++, and so I do not think I am qualified to become the
maintainer of this library.  The reason for the PR is that I simply wanted to
lend a helping hand with the CVEs.

File: pkgs/development/libraries/cereal/default.nix [hunk 1]
```diff
@@ -12,7 +12,20 @@ stdenv.mkDerivation rec {
...
-  cmakeFlagsArray = [ "-DJUST_INSTALL_CEREAL=yes" ];
+  patches = [
+    # https://nvd.nist.gov/vuln/detail/CVE-2020-11105
+    # serialized std::shared_ptr variables cannot always be expected to
+    # serialize back into their original values. This can have any number of
+    # consequences, depending on the context within which this manifests.
+    (fetchpatch {
+      name = "CVE-2020-11105.patch";
+      url =
+        "https://github.com/USCiLab/cereal/commit/3e8b9cab82dc4a596dc3bcfe7437946395ad782b.patch";
```
Comment (SuperSandro2000) [inline pkgs/development/libraries/cereal/default.nix:23]:
```suggestion
      url = "https://github.com/USCiLab/cereal/commit/3e8b9cab82dc4a596dc3bcfe7437946395ad782b.patch";
```

Comment (talw) [reply pkgs/development/libraries/cereal/default.nix]: Thanks!

It is probably a `nixfmt` quirk due to line length limit.

Applied your suggestion.

**Review by delroth — APPROVED** [review]

**Pull request merged**